File: 05Feb2025/backend_py/routes/members.py
"""
Module contains the routes for the members API.

Endpoints:
    - GET /members: Get all members.
    - GET /members/{member_id}: Get a member by ID.
    - POST /members: Add a new member.
    - PUT /members/{member_id}: Update a member by ID.
    - DELETE /members/{member_id}: Delete a member by ID.
"""

from fastapi import APIRouter, HTTPException
from fastapi.encoders import jsonable_encoder
from models import Member
from database import loadData, saveData, MEMBERS_FILE

router = APIRouter(prefix="/members", tags=["Members"])

@router.get("/")
def getMembers() -> list:
    """
    Get all members.
    
    Returns:
        list: The members data.
    """
    members = loadData(MEMBERS_FILE)
    return members["members"]

@router.get("/{member_id}")
def getMember(member_id: int) -> dict:
    """
    Get a member by ID.
    
    Args:
        member_id (int): The ID of the member to retrieve.
    
    Returns:
        dict: The member details.
    
    Raises:
        HTTPException: If the member is not found.
    """
    members = loadData(MEMBERS_FILE)
    member = next((m for m in members["members"] if m["id"] == member_id), None)

    if not member:
        raise HTTPException(status_code=404, detail="Member not found")

    return member

@router.post("/")
def addMember(member_data: Member) -> dict:
    """
    Add a new member.
    
    Args:
        member_data (Member): The member details to add.
    
    Returns:
        dict: The member details.
    """
    members = loadData(MEMBERS_FILE)

    # Generate new member ID
    new_member_id = members["members"][-1]["id"] + 1 if members["members"] else 1
    member_data.id = new_member_id
    new_member = jsonable_encoder(member_data)

    members["members"].append(new_member)
    saveData(MEMBERS_FILE, members)

    return new_member

@router.put("/{member_id}")
def updateMember(member_id: int, memberData: Member) -> dict:
    """
    Update a member by ID.
    
    Args:
        member_id (int): The ID of the member to update.
        memberData (Member): The updated member details.
    
    Returns:
        dict: Success message.
    
    Raises:
        HTTPException: If the member is not found.
    """
    members = loadData(MEMBERS_FILE)
    member = next((m for m in members["members"] if m["id"] == member_id), None)

    if not member:
        raise HTTPException(status_code=404, detail="Member not found")

    member["name"] = memberData.name
    saveData(MEMBERS_FILE, members)

    return {"message": "Member updated successfully"}

@router.delete("/{member_id}")
def deleteMember(member_id: int) -> dict:
    """
    Delete a member by ID.
    
    Args:
        member_id (int): The ID of the member to delete.
    
    Returns:
        dict: Success message.
    
    Raises:
        HTTPException: If the member is not found.
    """
    members = loadData(MEMBERS_FILE)
    member = next((m for m in members["members"] if m["id"] == member_id), None)

    if not member:
        raise HTTPException(status_code=404, detail="Member not found")

    members["members"].remove(member)
    saveData(MEMBERS_FILE, members)

    return {"message": "Member deleted successfully"}

File: LibraryManagement2/backend_js/routes/members.js
/**
 * Express router providing member related routes.
 * @module routes/members
 * @requires express.Router()
 * @requires file_handling
 * @exports router
 */

const express = require('express');
const router = express.Router();
const { loadFile, saveFile, MEMBERS_FILE } = require('../file_handling');

/**
 * GET /members
 * Retrieves all members.
 * @name getMembers
 * @function
 * @param {Object} req - Express request object.
 * @param {Object} res - Express response object.
 */
router.get('/', async (req, res) => {
    try {
        const data = await loadFile(MEMBERS_FILE);
        res.json(data.members);
    } catch (error) {
        res.status(500).json({ error: 'Failed to load members' });
    }
});

/**
 * GET /members/:id
 * Retrieves a member by ID.
 * @name getMemberById
 * @function
 * @param {Object} req - Express request object.
 * @param {Object} res - Express response object.
 */
router.get('/:id', async (req, res) => {
    try {
        const data = await loadFile(MEMBERS_FILE);
        const member = data.members.find(member => member.id === parseInt(req.params.id));
        if (!member) {
            return res.status(404).json({ error: 'Member not found' });
        }
        res.json(member);
    } catch (error) {
        res.status(500).json({ error: 'Failed to load member' });
    }
});

/**
 * POST /members
 * Adds a new member.
 * @name addMember
 * @function
 * @param {Object} req - Express request object.
 * @param {Object} res - Express response object.
 */
router.post('/', async (req, res) => {
    try {
        const { name } = req.body;
        
        if (!name) {
            return res.status(400).json({ error: 'Name is required' });
        }

        const data = await loadFile(MEMBERS_FILE);
        
        // Generate new ID
        const maxId = data.members.reduce((max, member) => Math.max(max, member.id), 0);
        const newMember = {
            id: maxId + 1,
            name
        };

        data.members.push(newMember);
        await saveFile(MEMBERS_FILE, data);
        res.status(201).json(newMember);
    } catch (error) {
        res.status(500).json({ error: 'Failed to add member' });
    }
});

/**
 * PUT /members/:id
 * Updates a member by ID.
 * @name updateMember
 * @function
 * @param {Object} req - Express request object.
 * @param {Object} res - Express response object.
 */
router.put('/:id', async (req, res) => {
    try {
        const { name } = req.body;
        const data = await loadFile(MEMBERS_FILE);
        const memberIndex = data.members.findIndex(member => member.id === parseInt(req.params.id));
        
        if (memberIndex === -1) {
            return res.status(404).json({ error: 'Member not found' });
        }

        data.members[memberIndex] = {
            ...data.members[memberIndex],
            name: name || data.members[memberIndex].name
        };

        await saveFile(MEMBERS_FILE, data);
        res.json(data.members[memberIndex]);
    } catch (error) {
        res.status(500).json({ error: 'Failed to update member' });
    }
});

/**
 * DELETE /members/:id
 * Deletes a member by ID.
 * @name deleteMember
 * @function
 * @param {Object} req - Express request object.
 * @param {Object} res - Express response object.
 */
router.delete('/:id', async (req, res) => {
    try {
        const data = await loadFile(MEMBERS_FILE);
        const memberIndex = data.members.findIndex(member => member.id === parseInt(req.params.id));
        
        if (memberIndex === -1) {
            return res.status(404).json({ error: 'Member not found' });
        }

        data.members.splice(memberIndex, 1);
        await saveFile(MEMBERS_FILE, data);
        res.json({ message: 'Member deleted successfully' });
    } catch (error) {
        res.status(500).json({ error: 'Failed to delete member' });
    }
});

module.exports = router;